x/gamm/pool-models/balancer/amm.go

```diff
@@ -282,34 +284,34 @@ func (p *Pool) CalcJoinPoolShares(_ sdk.Context, tokensIn sdk.Coins, swapFee sdk
 	}
 
 	// Add all exact coins we can (no swap). ctx arg doesn't matter for Balancer.
-	numShares, remCoins, err := cfmm_common.MaximalExactRatioJoin(p, sdk.Context{}, tokensIn)
+	numShares, remainingTokensIn, err := cfmm_common.MaximalExactRatioJoin(p, ctx, tokensIn)
 	if err != nil {
 		return sdk.ZeroInt(), sdk.NewCoins(), err
 	}
 
-	// update liquidity for accurate calcSingleAssetJoin calculation
-	newLiquidity = tokensIn.Sub(remCoins)
-	for _, coin := range newLiquidity {
-		poolAsset := poolAssetsByDenom[coin.Denom]
-		poolAsset.Token.Amount = poolAssetsByDenom[coin.Denom].Token.Amount.Add(coin.Amount)
-		poolAssetsByDenom[coin.Denom] = poolAsset
-	}
-
-	newTotalShares := totalShares.Add(numShares)
+	// This is new liquidity that to be added to the pool from exact ratio join.
+	// If there are remainingTokensIn, we will add it to newLiquidity by
+	// performing single asset join on each token in remainingTokensIn.
+	newLiquidity = tokensIn.Sub(remainingTokensIn)
 
-	// If there are coins that couldn't be perfectly joined, do single asset joins
+	// If there are tokens that couldn't be perfectly joined, do single asset joins
 	// for each of them.
-	if !remCoins.Empty() {
-		for _, coin := range remCoins {
-			newShares, err := p.calcSingleAssetJoin(coin, swapFee, poolAssetsByDenom[coin.Denom], newTotalShares)
-			if err != nil {
-				return sdk.ZeroInt(), sdk.NewCoins(), err
-			}
-
-			newLiquidity = newLiquidity.Add(coin)
-			newTotalShares = newTotalShares.Add(newShares)
-			numShares = numShares.Add(newShares)
+	if !remainingTokensIn.Empty() {
+		// update pool assets for accurate calcSingleAssetJoin calculation
```

Comment:
Doesn't calling `MaximalExactRatioJoin` already update the pool? Why do we need to add assets into the pool in a separate function?

Reply:
AFAIK, `MaximalExactRatioJoin` shouldnt be updating pool itself, just calculate the maximal amount of share available for join

Reply:
Yes, I believe @mattverse is correct here